src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -2873,6 +2873,322 @@ static void Main()
+            //     static void M2([Optional, DecimalConstant(1, 0, 0u, 0u, 11u)] ref readonly decimal d = 1.1m) => throw null;
+            Diagnostic(ErrorCode.WRN_RefReadonlyParameterDefaultValue, "1.1m").WithArguments("d").WithLocation(6, 92),
+            // (9,9): error CS7036: There is no argument given that corresponds to the required parameter 'd' of 'C.M1(ref readonly decimal)'
+            //         M1();
+            Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "M1").WithArguments("d", "C.M1(ref readonly decimal)").WithLocation(9, 9));
+    }
+
+    [Fact]
+    public void DefaultParameterValue_DateTimeConstant_Valid()
+    {
+        var source = """
+            using System;
+            using System.Runtime.CompilerServices;
+            using System.Runtime.InteropServices;
+            class C
+            {
+                static void M([Optional, DateTimeConstant(100L)] ref readonly DateTime d) => Console.Write(d.Ticks);
+                static void Main() => M();
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "100").VerifyDiagnostics(
+            // (6,30): warning CS9521: A default value is specified for 'ref readonly' parameter 'd', but 'ref readonly' should be used only for references. Consider declaring the parameter as 'in'.
+            //     static void M([Optional, DateTimeConstant(100L)] ref readonly DateTime d) => Console.Write(d.Ticks);
+            Diagnostic(ErrorCode.WRN_RefReadonlyParameterDefaultValue, "DateTimeConstant(100L)").WithArguments("d").WithLocation(6, 30));
+        verifier.VerifyIL("C.Main", """
+            {
+              // Code size       17 (0x11)
+              .maxstack  1
+              .locals init (System.DateTime V_0)
+              IL_0000:  ldc.i4.s   100
+              IL_0002:  conv.i8
+              IL_0003:  newobj     "System.DateTime..ctor(long)"
+              IL_0008:  stloc.0
+              IL_0009:  ldloca.s   V_0
+              IL_000b:  call       "void C.M(ref readonly System.DateTime)"
+              IL_0010:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void DefaultParameterValue_DateTimeConstant_Invalid()
+    {
+        var source = """
+            using System;
+            using System.Runtime.CompilerServices;
+            class C
+            {
+                static void M([DateTimeConstant(100L)] ref readonly DateTime d) => throw null;
+                static void Main()
+                {
+                    M();
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (8,9): error CS7036: There is no argument given that corresponds to the required parameter 'd' of 'C.M(ref readonly DateTime)'
+            //         M();
+            Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "M").WithArguments("d", "C.M(ref readonly System.DateTime)").WithLocation(8, 9));
+    }
```

Comment:
> } [](http://example.com/codeflow?start=3&length=2)

It looks like we are not testing consumption of default values from metadata. I think we should cover at least all valid forms of encoding: attributes only, constant value, combination of both.